# MULTIPLE REGRESSION

$The \ general \ additive \ multiple \ regression \ model \ equation \ is $

$$ y = \beta_0 +\beta_1 x_1 + \beta_2 x_2 + \beta_3 x_3 + \epsilon$$

$where \ E(\epsilon)= 0 \ and \ V(\epsilon) = \sigma^2. In \ addition \ it \ is \ assumed \ that \ the \ error\  term \epsilon \ follows \ normal \ distribution $

### $Adjusted R^2$

In multiple linear regression value od $R^2$ can be inflated by adding lots of independent variables even if most of them are rather unusefull. 

So objective of multiple regression is not simply explain the variation in $y$, but also do so using a model with relatively few independent variables that are easily interpreted. It is thus desirable to $adjust R^2$ 

$$Adjusted R^2 = 1 - \frac{(1-R^2)(n-1)}{n-k-1}$$

Where $$n = number \ of \ sample \ data$$ and $$k = number \ of \ predictors \ or \ independent \ variables$$

How to use $Adjusted R^2$
- Larger the number of predictors $k$ relative to the sample size $n$, the smaller $Adjusted R^2$ will  be relative to $R^2$. So if $Adjusted R^2$ is substancially smaller than $R^2$, it indicates the model may contains too many predictors.
- The $Adjusted\ R^2$ value will increase only when the addition of a new independent variable improves the model's fit, and it will decrease when the new variable does not contribute significantly to the model's fit. This adjustment helps in selecting the most relevant independent variables for the regression model.

### Model Utility test

With multivariate data, there is no picture analogous to a scatter plot to indicate
whether a particular multiple regression model will successfully explain observed
y variation. The value of $R^2$ certainly communicates a preliminary message, but this
value is sometimes deceptive because it can be greatly inflated by using a large
number of predictors relative to the sample size. For this reason, it is important to
have a formal test for model utility.


The model utility test in simple linear regression involved the null hypothesis
$H_0 : \beta_1 = 0$, according to which there is no useful relation between y and the single
predictor x.


In multiple linear regrssion we consider the assertion that $\beta_1 = 0, \beta_2 = 0, ...., \beta_k =0$ which
says that there is no useful relationship between y and any of the k predictors

Test           : F test

$H_0 : \beta_1 = \beta_2 = \beta_3 = ... \beta_k = 0$

$H_a : one \ of \ the\ \beta_i  \ is \ non \ zero$

Test statistic : $f = \frac{R^2/k}{(1-R^2)(n-k-1)}$


Rejection Region : Upper tail(Right tailed)

---------
----------
---------

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f

## EXAMPLE 1

The article “How to Optimize and Control the Wire Bonding Process: Part II” (Solid
State Technology, Jan. 1991: 67–72) described an experiment carried out to assess the
impact of the variables , $x_1 = force (gm)$, $x_2 = power(mW)$, $x_3 = temperature(^0 C)$
and $x_4 = time(msec)$ on$y = $ bond shear strength (gm). The following data was
generated to be consistent with the information given in the article:

In [2]:
data = [
    [30, 60, 175, 15, 26.2],
    [40, 60, 175, 15, 26.3],
    [30, 90, 175, 15, 39.8],
    [40, 90, 175, 15, 39.7],
    [30, 60, 225, 15, 38.6],
    [40, 60, 225, 15, 35.5],
    [30, 90, 225, 15, 48.8],
    [40, 90, 225, 15, 37.8],
    [30, 60, 175, 25, 26.6],
    [40, 60, 175, 25, 23.4],
    [30, 90, 175, 25, 38.6],
    [40, 90, 175, 25, 52.1],
    [30, 60, 225, 25, 39.5],
    [40, 60, 225, 25, 32.3],
    [30, 90, 225, 25, 43.0],
    [40, 90, 225, 25, 56.0],
    [25, 75, 200, 20, 35.2],
    [45, 75, 200, 20, 46.9],
    [35, 45, 200, 20, 22.7],
    [35, 105, 200, 20, 58.7],
    [35, 75, 150, 20, 34.5],
    [35, 75, 250, 20, 44.0],
    [35, 75, 200, 10, 35.7],
    [35, 75, 200, 30, 41.8],
    [35, 75, 200, 20, 36.5],
    [35, 75, 200, 20, 37.6],
    [35, 75, 200, 20, 40.3],
    [35, 75, 200, 20, 46.0],
    [35, 75, 200, 20, 27.8],
    [35, 75, 200, 20, 40.3]
       ]

df = pd.DataFrame(data, columns=[
    'Force', 'Power', 'Temperature', 'Time', 'Strength'])

df.head()

Force  Power  Temperature  Time  Strength
0     30     60          175    15      26.2
1     40     60          175    15      26.3
2     30     90          175    15      39.8
3     40     90          175    15      39.7
4     30     60          225    15      38.6

- Calculate an estimated regression line
- Calculate $R^2$ value for the moodel
- Calculate $Adusted R^2$ value for the model
- Do model utility test

------------

## SOLUTION

### Fit model

Data is about impact of Force, Power, Temperature,Time on Strength.

So indenpendent variables  Force, Power, Temperature,Time and dependent variable is Strength

In [3]:
X = df[['Force', 'Power', 'Temperature', 'Time']]
y = df['Strength']

In [4]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()

In [5]:
regressor.fit(X,y)

LinearRegression()

In [6]:
# Coefficients
print("Intercept:", regressor.intercept_)
print("Coefficients:", regressor.coef_)

Intercept: -37.47666666666663
Coefficients: [0.21166667 0.49833333 0.12966667 0.25833333]


### Estimated regression line 
$$ y = -37.4766 + 0.2116 x_1 + 0.4983 x_2 + 0.1296 x_3 +  0.2583 x_4 $$

### Predictions, Error and Standardized error

In [7]:
predicted_y = regressor.predict(X)

In [8]:
error = np.array(y - predicted_y)

In [9]:
df['error'] = error

In [10]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

df['standardized_error'] = scaler.fit_transform(df[['error']])

### $R^2$ Value

In [11]:
r_square = regressor.score(X,y)

In [12]:
r_square

0.7139592785089428

### Adjsuted $R^2$

In [13]:
# To Calculate Adjusted R^2

n = df.shape[0]      #number of observations

k = X.shape[1]         #numbeer of independent variables

adj_r_square = 1 - ((1- r_square)*(n-1))/(n - k - 1)

print("Adjusted R_square ", adj_r_square)

Adjusted R_square  0.6681927630703737


 The value of $R^2$, and $Adjusted R^2$ certainly suggest a useful model.

### Model Utility test

In [14]:
# To define F statistic
# n and k already calculated for finding Adjusted R^2

f_stat = (r_square/k)/ (1-r_square)*(n-k-1)

In [15]:
P_value = 1 - f.cdf(f_stat, k, n-k-1)  #P value for right tailed test

In [16]:
P_value

1.5924160694513745e-06

P_value is less than 0.01, so we can reject $H_0$, and can conclude there is  a useful linear relationship between y and at least one of the four predictors in the model.

Note: This does not mean that all four predictors are usefull

-------------
----------------

## EXAMPLE 2

Investigators carried out a study to see how various characteristics of concrete are influenced by $x_1 =$ % limestone powder and $x_2 = $ water-cement ratio, resulting in the
accompanying data (“Durability of Concrete with Addition of Limestone Powder,”
Magazine of Concrete Research, 1996: 131–137).

In [17]:
data = {
    'x1': [21, 21, 7,7, 28, 0, 14, 14,14],
    'x2': [0.65, 0.55, 0.65, 0.55, 0.60, 0.60, 0.70, 0.50, 0.60],
    'x1*x2': [13.65, 11.55, 4.55,3.85, 16.80, 0.00, 9.80, 7.00,8.40],
    '28-day Comp Str (MPa)': [33.55, 47.55, 35.00, 35.90,40.90, 39.10, 31.55, 48.00,42.30],
    'Adsorbability (%)': [8.42, 6.26, 6.74, 6.59,7.28, 6.90, 10.80, 5.63,7.43]
}

# Create the DataFrame
df_2 = pd.DataFrame(data)

In [18]:
df_2

x1    x2  x1*x2  28-day Comp Str (MPa)  Adsorbability (%)
0  21  0.65  13.65                  33.55               8.42
1  21  0.55  11.55                  47.55               6.26
2   7  0.65   4.55                  35.00               6.74
3   7  0.55   3.85                  35.90               6.59
4  28  0.60  16.80                  40.90               7.28
5   0  0.60   0.00                  39.10               6.90
6  14  0.70   9.80                  31.55              10.80
7  14  0.50   7.00                  48.00               5.63
8  14  0.60   8.40                  42.30               7.43

- ANALYSIS 1
    - Fit the linear regression model with dependent variable as strength and independent variables x1 and x2
    - Calculate $R^2$ and $ Adjusted R^2$
- ANALYSIS 2
    - Fit the linear regression model with dependent variable as strength and independent variables x1, x2 and the     interaction predictor x1x2
    - Calculate $R^2$ and $ Adjusted R^2$
- ANALYSIS 3
        - Fit linear equation with all the variables. Find $R^2$ and adjusted $R^2$
        
- Compare above values of $Adjusted \ R^2$ and say which model is better

#### ANALYSIS 1

In [19]:
X  = df_2[['x1','x2']]
y = df_2['28-day Comp Str (MPa)']

In [20]:
regressor.fit(X,y)     

LinearRegression()

In [21]:
# Coefficients
print("Intercept:", regressor.intercept_)
print("Coefficients:", regressor.coef_)

Intercept: 84.81666666666669
Coefficients: [  0.16428571 -79.66666667]


$ Equation \ of \ regression \ line \ is $

$$y = 84.8166 + 0.1642 x_1 + -79.6666 x_2$$

In [22]:
r_square = regressor.score(X,y)
r_square

0.7405918080363822

In [23]:
# To Calculate Adjusted R^2

n = df_2.shape[0]      #number of observations

k = X.shape[1]         #numbeer of independent variables

adj_r_square = 1 - ((1- r_square)*(n-1))/(n - k - 1)

print("Adjusted R_square ", adj_r_square)

Adjusted R_square  0.6541224107151764


#### ANALYSIS 2

In [24]:
X  = df_2[['x1','x2','x1*x2']]
y = df_2['28-day Comp Str (MPa)']

In [25]:
regressor.fit(X,y)

LinearRegression()

In [26]:
# Coefficients
print("Intercept:", regressor.intercept_)
print("Coefficients:", regressor.coef_)

Intercept: 6.2166666666663914
Coefficients: [ 5.77857143 51.33333333 -9.35714286]


$Equation \ for \  the \ regression \ line \ is $

$$ y = 6.2166 + 5.7785 x1 + 51.3333x2 + -9.3571 x1 x2 $$

In [27]:
r_square = regressor.score(X,y)
r_square

0.8946264548364897

In [28]:
# To Calculate Adjusted R^2

n = df_2.shape[0]      #number of observations

k = X.shape[1]         #numbeer of independent variables

adj_r_square = 1 - ((1- r_square)*(n-1))/(n - k - 1)

print("Adjusted R_square ", adj_r_square)

Adjusted R_square  0.8314023277383835


$R^2$ and $Adjusted R^2$ value incresed when we added interaction predictor $x1x2$

So second model is much better than the first one

#### ANALYSIS 2

In [29]:
X  = df_2[['x1','x2','x1*x2','Adsorbability (%)' ]]
y = df_2['28-day Comp Str (MPa)']

In [30]:
regressor.fit(X,y)

LinearRegression()

In [31]:
r_square = regressor.score(X,y)
r_square

0.9042623255458607

In [32]:
# To Calculate Adjusted R^2

n = df_2.shape[0]      #number of observations

k = X.shape[1]         #numbeer of independent variables

adj_r_square = 1 - ((1- r_square)*(n-1))/(n - k - 1)

print("Adjusted R_square ", adj_r_square)

Adjusted R_square  0.8085246510917214


There is not much change in the values of $R^2$ and $Adjusted R^2$ from 2nd model.